server/src/test/java/org/elasticsearch/cluster/NodeConnectionsServiceTests.java

```diff
@@ -245,13 +249,111 @@ public String toString() {
     }
 
+    public void testDisconnectionHistory() {
+        final Settings.Builder settings = Settings.builder();
+        settings.put(CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.getKey(), "100ms");
+
+        final DeterministicTaskQueue deterministicTaskQueue = new DeterministicTaskQueue();
+        final ThreadPool threadPool = deterministicTaskQueue.getThreadPool();
+
+        MockTransport transport = new MockTransport(deterministicTaskQueue.getThreadPool());
+        TestTransportService transportService = new TestTransportService(transport, deterministicTaskQueue.getThreadPool());
+        transportService.start();
+        transportService.acceptIncomingRequests();
+
+        final NodeConnectionsService service = new NodeConnectionsService(settings.build(), threadPool, transportService);
+        service.start();
+
+        final DiscoveryNode noClose = DiscoveryNodeUtils.create("noClose");
+        final DiscoveryNode gracefulClose = DiscoveryNodeUtils.create("gracefulClose");
+        final DiscoveryNode exceptionalClose = DiscoveryNodeUtils.create("exceptionalClose");
+
+        nodeCloseExceptions.put(exceptionalClose, new RuntimeException());
+
+        final AtomicBoolean connectionCompleted = new AtomicBoolean();
+        DiscoveryNodes nodes = DiscoveryNodes.builder().add(noClose).add(gracefulClose).add(exceptionalClose).build();
+
+        service.connectToNodes(nodes, () -> connectionCompleted.set(true));
+        deterministicTaskQueue.runAllRunnableTasks();
+        assertTrue(connectionCompleted.get());
+
+        assertNullDisconnectionHistory(service, noClose);
+        assertNullDisconnectionHistory(service, gracefulClose);
+        assertNullDisconnectionHistory(service, exceptionalClose);
+
+        try (var mockLog = MockLog.capture(NodeConnectionsService.class)) {
+            mockLog.addExpectation(
+                new MockLog.SeenEventExpectation(
+                    "reconnect after graceful close",
+                    NodeConnectionsService.class.getCanonicalName(),
+                    Level.WARN,
+                    "reopened transport connection to node ["
+                        + gracefulClose.descriptionWithoutAttributes()
+                        + "] which disconnected gracefully [*ms] ago "
+                        + "but did not restart, so the disconnection is unexpected; "
+                        + "see [https://www.elastic.co/docs/*] for troubleshooting guidance"
+                )
+            );
+            mockLog.addExpectation(
+                new MockLog.SeenEventExpectation(
+                    "reconnect after exceptional close",
+                    NodeConnectionsService.class.getCanonicalName(),
+                    Level.WARN,
+                    "reopened transport connection to node ["
+                        + exceptionalClose.descriptionWithoutAttributes()
+                        + "] which disconnected exceptionally [*ms] ago "
```

Comment:
We control the passage of time so we can assert that the reconnect happens exactly one reconnect interval later:

```suggestion
                        + "] which disconnected exceptionally ["
                        + CLUSTER_NODE_RECONNECT_INTERVAL_SETTING.get(Settings.EMPTY).millis()
                        + "ms] ago "
```